File: src/systems/screen_rect.cpp
#include "screen_rect.hpp"
#include "../ecsql/system.hpp"
#include "raylib.h"

void register_update_screen_rect(ecsql::World& world) {
	world.register_system({
		"UpdateScreenRect",
		R"(
			INSERT OR REPLACE INTO Rectangle(entity_id, width, height)
			SELECT entity_id, width, height
			FROM ScreenRect
				JOIN screen_size
		)",
		[](auto& sql) {
			static bool first_time = true;
			if (first_time || IsWindowResized()) {
				sql();
				first_time = false;
			}
		},
	});
}

File: assets/components/screen_rect.lua
component "ScreenRect" {
    additional_schema = [[
        CREATE TRIGGER ScreenRect_inserted
        AFTER INSERT ON ScreenRect
        BEGIN
            INSERT OR REPLACE INTO Rectangle(entity_id, width, height)
            SELECT new.entity_id, width, height
            FROM screen;
        END;

        CREATE TRIGGER ScreenRect_screen_updated
        AFTER UPDATE OF width, height ON screen
        BEGIN
            UPDATE Rectangle
            SET width = s.width, height = s.height
            FROM (
                SELECT entity_id, width, height
                FROM ScreenRect, screen
            ) AS s
            WHERE Rectangle.entity_id = s.entity_id;
        END;
    ]]
}
